src/core/server/elasticsearch/client/client_config.ts

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { ConnectionOptions as TlsConnectionOptions } from 'tls';
+import { URL } from 'url';
+import { Duration } from 'moment';
+import { ClientOptions, NodeOptions } from '@elastic/elasticsearch';
+import { ElasticsearchConfig } from '../elasticsearch_config';
+
+/**
+ * @privateRemarks Config that consumers can pass to the Elasticsearch JS client is complex and includes
+ * not only entries from standard `elasticsearch.*` yaml config, but also some Elasticsearch JS
+ * client specific options like `keepAlive` or `plugins` (that eventually will be deprecated).
+ *
+ * @public
+ */
+export type ElasticsearchClientConfig = Pick<
+  ElasticsearchConfig,
+  | 'customHeaders'
+  | 'logQueries'
+  | 'sniffOnStart'
+  | 'sniffOnConnectionFault'
+  | 'requestHeadersWhitelist'
+  | 'sniffInterval'
+  | 'hosts'
+  | 'username'
+  | 'password'
+> & {
+  pingTimeout?: ElasticsearchConfig['pingTimeout'] | ClientOptions['pingTimeout'];
+  requestTimeout?: ElasticsearchConfig['requestTimeout'] | ClientOptions['requestTimeout'];
+  ssl?: Partial<ElasticsearchConfig['ssl']>;
+};
+
+export function parseClientOptions(
+  config: ElasticsearchClientConfig,
+  scoped: boolean
+): ClientOptions {
+  const clientOptions: ClientOptions = {
+    sniffOnStart: config.sniffOnStart,
+    sniffOnConnectionFault: config.sniffOnConnectionFault,
+    headers: config.customHeaders,
```

Comment:
This is a pretty straightforward adaptation of `src/core/server/elasticsearch/elasticsearch_client_config.ts`.

Notable changes:
- `apiVersion` `keepAlive` `log` and `plugins` options are no longer used and were deleted from `ElasticsearchClientConfig`
- `config.customHeaders` are now injected on the root of the client options instead of in each `host`/`node`
- I removed the `ElasticsearchClientConfigOverrides` options in favor of a single `scoped` boolean, as this object was only used to override behavior for scoped clients.

Reply:
Hello! Just a nit, `keepAlive`, `keepAliveInterval`, `keepAliveMaxFreeSockets`, and `keepAliveFreeSocketTimeout` have been removed (I'll update the breaking changes doc), now you can configure the keep-alive agent with the [`agent`](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/client-configuration.html) option, which accepts the same parameters as the Node.js core one.

Reply:
Oh thanks, totally missed the `ClientOptions.agent.keepAlive`! Will add the option back to our wrapper.